frontend/lib/src/components/widgets/DataFrame/hooks/useSelectionHandler.ts

```diff
@@ -0,0 +1,195 @@
+
+  const isColumnSelectionActivated =
+    !isEmptyTable &&
+    !isDisabled &&
+    (element.selectionMode.includes(ArrowProto.SelectionMode.SINGLE_COLUMN) ||
+      element.selectionMode.includes(ArrowProto.SelectionMode.MULTI_COLUMN))
+  const isMultiColumnSelectionActivated =
+    isColumnSelectionActivated &&
+    element.selectionMode.includes(ArrowProto.SelectionMode.MULTI_COLUMN)
+
+  const isRowSelected = gridSelection.rows.length > 0
+  const isColumnSelected = gridSelection.columns.length > 0
+  const isCellSelected = gridSelection.current !== undefined
+
+  const processSelectionChange = React.useCallback(
+    (newSelection: GridSelection) => {
+      const rowSelectionChanged = !isEqual(
+        newSelection.rows.toArray(),
+        gridSelection.rows.toArray()
+      )
+
+      const columnSelectionChanged = !isEqual(
+        newSelection.columns.toArray(),
+        gridSelection.columns.toArray()
+      )
+
+      // A flag to determine if the selection should be synced with the widget state
+      let syncSelection =
+        (isRowSelectionActivated && rowSelectionChanged) ||
+        (isColumnSelectionActivated && columnSelectionChanged)
+
+      let updatedSelection = newSelection
+      if (
+        (isRowSelectionActivated || isColumnSelectionActivated) &&
+        newSelection.current !== undefined
+      ) {
+        // The default behavior is that row selections are cleared when a cell is selected.
+        // This is not desired when row selection is activated. Instead, we want to keep the
+        // row selection and only update the cell selection.
+        updatedSelection = {
+          ...newSelection,
+          rows: gridSelection.rows,
+          columns: gridSelection.columns,
+        }
+        // It should not sync the selection
+        // when only the cell selection changes
+        syncSelection = false
+      }
+
+      if (
+        rowSelectionChanged &&
+        newSelection.rows.length > 0 &&
+        columnSelectionChanged &&
+        newSelection.columns.length === 0
+      ) {
+        // Keep the column selection if row selection was changed
+        updatedSelection = {
+          ...updatedSelection,
+          columns: gridSelection.columns,
+        }
+        syncSelection = true
+      }
+      if (
+        columnSelectionChanged &&
+        newSelection.columns.length > 0 &&
+        rowSelectionChanged &&
+        newSelection.rows.length === 0
+      ) {
+        // Keep the row selection if column selection was changed
+        updatedSelection = {
+          ...updatedSelection,
+          rows: gridSelection.rows,
+        }
+        syncSelection = true
+      }
+
+      setGridSelection(updatedSelection)
+
+      if (syncSelection) {
+        syncSelectionState(updatedSelection)
+      }
+    },
+    [
+      gridSelection,
+      isRowSelectionActivated,
+      isColumnSelectionActivated,
+      syncSelectionState,
```

Comment:
do we have to pass `setGridSelection` here, too? It's a state-setter so probably not as it usually does not change, but want to sanity check

Reply:
Nope, afaik state-setter methods don't need to be included in dependencies and eslint also seems to be fine with this.